server/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -219,7 +383,29 @@ boolean compilationLimitsEnabled() {
 
     void registerClusterSettingsListeners(ClusterSettings clusterSettings) {
         clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_SIZE_IN_BYTES, this::setMaxSizeInBytes);
-        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_RATE, compiler::setMaxCompilationRate);
+        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED,
+                // Don't deref potentially null generalCache
+                r -> {
+                    if (generalCache != null) {
+                        generalCache.setMaxCompilationRate(r);
+                    }
+                },
+                s -> {
```

Comment:
I think we need to consider changing the signature here to allow the old value to be passed through. We have it in the update consumer within Setting. Also, using the cross setting validation should allow simpler validation here I think. See the other validate methods that can be implemented on Setting.Validator.

Reply:
We've taken a step back and are changing the semantics here as mentioned above.